File: src/net/detunized/unfuckjs/UnfuckJsIntention.java
package net.detunized.unfuckjs;

import com.intellij.codeInsight.intention.PsiElementBaseIntentionAction;
import com.intellij.lang.javascript.JSElementTypes;
import com.intellij.lang.javascript.psi.*;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.util.IncorrectOperationException;
import org.intellij.idea.lang.javascript.psiutil.BinaryOperatorUtils;
import org.intellij.idea.lang.javascript.psiutil.JSElementFactory;
import org.jetbrains.annotations.NotNull;

public class UnfuckJsIntention extends PsiElementBaseIntentionAction {
    @Override
    public void invoke(@NotNull Project project, Editor editor, @NotNull PsiElement psiElement) throws IncorrectOperationException {
        JSStatement statement = PsiTreeUtil.getParentOfType(psiElement, JSStatement.class);
        if (statement != null)
            unfuckStatement(statement);
    }

    private void unfuckStatement(JSStatement s) {
        if (s instanceof JSExpressionStatement)
            unfuckExpressionStatement((JSExpressionStatement)s);
        else if (s instanceof JSReturnStatement) {
            unfuckReturnStatement((JSReturnStatement)s);
        }
    }

    private void unfuckExpressionStatement(JSExpressionStatement s) {
        JSExpression e = s.getExpression();
        if (e instanceof JSCommaExpression) {
            splitComma(s, (JSCommaExpression)e);
        } else if (e instanceof JSBinaryExpression) {
            JSBinaryExpression be = (JSBinaryExpression)e;
            String op = BinaryOperatorUtils.getOperatorText(be.getOperationSign());
            if (op.equals("&&")) {
                convertAndToIf(s, be);
            }
        }
    }

    private void unfuckReturnStatement(JSReturnStatement s) {
        JSExpression e = s.getExpression();
        if (e instanceof JSCommaExpression) {
            splitComma(s, (JSCommaExpression)e, "return");
        }
    }

    private void splitComma(JSStatement s, JSCommaExpression e) {
        splitComma(s, e, "");
    }

    // TODO: Add support for while, for and others
    // keyword: return, throw
    private void splitComma(JSStatement s, JSCommaExpression e, String keyword) {
        String lhs = e.getLOperand().getText() + ';';
        String rhs = appendSpace(keyword) + e.getROperand().getText() + ';';

        JSStatement parent = PsiTreeUtil.getParentOfType(s, JSStatement.class);
        if (parent instanceof JSIfStatement) {
            JSIfStatement ifStatement = (JSIfStatement)parent;
            if (s == ifStatement.getThen() || s == ifStatement.getElse()) {
                JSBlockStatement block = (JSBlockStatement)JSElementFactory.replaceStatement(s, '{' + lhs + rhs + '}');

                for (JSStatement i: block.getStatements())
                    unfuckStatement(i);
            }
        } else {
            JSStatement ns1 = JSElementFactory.replaceStatement(s, lhs);
            JSStatement ns2 = JSElementFactory.addStatementAfter(ns1, rhs);

            unfuckStatement(ns1);
            unfuckStatement(ns2);
        }
    }

    private void convertAndToIf(JSStatement s, JSBinaryExpression e) {
        JSIfStatement ifs = (JSIfStatement)JSElementFactory.replaceStatement(
                s, String.format(
                        "if (%s) { %s; }",
                        e.getLOperand().getText(),
                        e.getROperand().getText()));

        unfuckStatement(ifs.getThen());
    }

    private String appendSpace(String text) {
        return text.isEmpty() ? text : text + ' ';
    }

    @Override
    public boolean isAvailable(@NotNull Project project, Editor editor, @NotNull PsiElement psiElement) {
        return true;
    }

    @NotNull
    @Override
    public String getFamilyName() {
        return "Unfuck";
    }

    @NotNull
    @Override
    public String getText() {
        return "Unfuck some";
    }
}

File: src/net/detunized/unfuckjs/UnfuckJsIntention.kt
package net.detunized.unfuckjs

import com.intellij.codeInsight.intention.PsiElementBaseIntentionAction
import com.intellij.lang.javascript.psi.*
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElement
import com.intellij.psi.util.PsiTreeUtil
import org.intellij.idea.lang.javascript.psiutil.BinaryOperatorUtils
import org.intellij.idea.lang.javascript.psiutil.JSElementFactory
import org.intellij.idea.lang.javascript.psiutil.ParenthesesUtils
import org.intellij.idea.lang.javascript.psiutil.BoolUtils

public class UnfuckJsIntention: PsiElementBaseIntentionAction() {
    override fun invoke(project: Project, editor: Editor, psiElement: PsiElement) {
        val statement = PsiTreeUtil.getParentOfType(psiElement, javaClass<JSStatement>())
        if (statement != null)
            unfuckStatement(statement)
    }

    private fun unfuckStatement(s: JSStatement) {
        when (s) {
            is JSExpressionStatement -> unfuckExpressionStatement(s)
            is JSReturnStatement -> unfuckReturnStatement(s)
            is JSIfStatement -> unfuckIfStatement(s)
        }
    }

    private fun unfuckExpressionStatement(s: JSExpressionStatement) {
        val e = s.getExpression()
        when (e) {
            is JSCommaExpression -> splitComma(s, e)
            is JSBinaryExpression -> {
                val op = BinaryOperatorUtils.getOperatorText(e.getOperationSign())
                when (op) {
                    "&&" -> convertAndToIf(s, e)
                    "||" -> convertOrToIf(s, e)
                }
            }
            is JSParenthesizedExpression -> stripParens(s, e)
        }
    }

    private fun unfuckReturnStatement(s: JSReturnStatement) {
        val e = s.getExpression()
        if (e is JSCommaExpression) {
            splitComma(s, e, "return")
        }
    }

    fun unfuckIfStatement(s: JSIfStatement) {
        fun isInBlock(s: JSStatement) =
                s.getParent() is JSBlockStatement

        fun addBefore(anchor: JSStatement, code: String) {
            if (isInBlock(anchor))
                JSElementFactory.addStatementBefore(anchor, "${code};\n")
            else
                JSElementFactory.replaceElementWithStatement(anchor, "{ ${code};\n${s.getText()} }")
        }

        fun addBefore(anchor: JSStatement, e: PsiElement) =
                addBefore(anchor, e.getText())

        val e = s.getCondition()
        if (e is JSCommaExpression) {
            val lhs = e.getLOperand();
            val rhs = e.getROperand()
            JSElementFactory.replaceExpression(e, rhs.getText());
            addBefore(s, lhs)
        }
    }

    private fun splitComma(s: JSStatement, e: JSCommaExpression) {
        splitComma(s, e, "")
    }

    // TODO: Add support for while, for and others
    // keyword: return, throw
    private fun splitComma(s: JSStatement, e: JSCommaExpression, keyword: String) {
        val lhs = e.getLOperand().getText() + ';'
        val rhs = appendSpace(keyword) + e.getROperand().getText() + ';'

        val parent = PsiTreeUtil.getParentOfType<JSStatement>(s, javaClass<JSStatement>())
        if (parent is JSIfStatement) {
            if (s == parent.getThen() || s == parent.getElse()) {
                val block = JSElementFactory.replaceStatement(s, '{' + lhs + rhs + '}') as JSBlockStatement
                for (i in block.getStatements())
                    unfuckStatement(i)
            }
        } else {
            val ns1 = JSElementFactory.replaceStatement(s, lhs)
            val ns2 = JSElementFactory.addStatementAfter(ns1, rhs)

            unfuckStatement(ns1)
            unfuckStatement(ns2)
        }
    }

    private fun convertAndToIf(s: JSStatement, e: JSBinaryExpression) {
        val ifs = JSElementFactory.replaceStatement(
                s,
                "if (%s) { %s; }".format(e.getLOperand().getText(), e.getROperand().getText())
        ) as JSIfStatement

        unfuckStatement((ifs.getThen() as JSBlockStatement).getStatements()[0])
    }

    private fun convertOrToIf(s: JSStatement, e: JSBinaryExpression) {
        val ifs = JSElementFactory.replaceStatement(
                s,
                "if (%s) { %s; }".format(
                        BoolUtils.getNegatedExpressionText(e.getLOperand()),
                        e.getROperand().getText()
                )
        ) as JSIfStatement

        unfuckStatement((ifs.getThen() as JSBlockStatement).getStatements()[0])
    }

    private fun stripParens(s: JSStatement, e: JSParenthesizedExpression) {
        val ns = JSElementFactory.replaceStatement(s, ParenthesesUtils.stripParentheses(e).getText())

        unfuckStatement(ns)
    }

    private fun appendSpace(text: String): String {
        return if (text.isEmpty()) text else text + ' '
    }

    override fun isAvailable(project: Project, editor: Editor, psiElement: PsiElement): Boolean {
        return false
    }

    override fun getFamilyName(): String {
        return "Unfuck"
    }

    override fun getText(): String {
        return "Unfuck some"
    }
}
